src/main/java/org/opentripplanner/graph_builder/linking/SimpleStreetSplitter.java

```diff
@@ -404,9 +440,11 @@ private SplitterVertex split (StreetEdge edge, LinearLocation ll, boolean tempor
 
         if (destructiveSplitting) {
             // update indices of new edges
-            idx.insert(edges.first.getGeometry(), edges.first);
-            idx.insert(edges.second.getGeometry(), edges.second);
-
+            synchronized (this) {
+                // Note: Write operations are not synchronized in HashGridSpatialIndex, hence the lock.
+                idx.insert(edges.first.getGeometry(), edges.first);
```

Comment:
Is it correct to say that edges are only inserted in the index during destructive splitting (and not inserted for nondestructive splitting), and therefore only destructively split edges can be re-split, because nondestructively split edges won't be found in the index? I believe this was the intended behavior (so that edges for origin and destination points are not re-split by other threads). But I'm not sure it's stated anywhere in the Javadoc. To someone first approaching this code, I think the fact that the split is destructive or nondestructive (permanent or temporary) will seem independent of the fact that edges are indexed (and therefore re-splittable). So the fact that these two characteristics are coupled should be very visibly stated in the Javadoc.

Reply:
Added some comments here: https://github.com/opentripplanner/OpenTripPlanner/pull/2762/commits/cf74d8132a221812ce2ed114a13c3faffd7cd282